remote_attestation/rust/src/attestation.rs

```diff
@@ -15,238 +15,545 @@
 //
 
 use crate::{
-    crypto::{get_sha256, AeadEncryptor, EncryptorType, KeyNegotiator},
-    proto::{AttestationIdentity, AttestationInfo, AttestationReport},
+    crypto::{
+        get_random, get_sha256, AeadEncryptor, KeyNegotiator, KeyNegotiatorType, SignatureVerifier,
+        Signer,
+    },
+    proto::{AttestationInfo, AttestationInit, AttestationReport, ClientIdentity, ServerIdentity},
 };
 use anyhow::{anyhow, Context};
 
-/// Remote attestation protocol implementation.
-/// Template `S` defines the type of attestation used. It can be one of:
-/// - [`UnattestedSelf`]: represents an attestation process, where current machine remotely attests
-///   to a remote peer and sends attestation info to it.
-/// - [`UnattestedPeer`]: represents an attestation process, where current machine remotely attests
-///   a remote peer and verifies its attestation info.
-/// - [`MutuallyUnattested`]: represents an attestation process, where current machine and a remote
-///   peer remotely attest each other.
-pub struct AttestationEngine<S>
+const ATTESTATION_PROTOCOL_VERSION: u8 = 1;
+/// Size (in bytes) of a random array sent in messages to prevent replay attacks.
+const REPLAY_PROTECTION_ARRAY_SIZE_BYTES: usize = 32;
+
+/// Convenience trait for transitioning between states in the state machine.
+pub trait StateMachine<T, A> {
+    fn transition(self, argument: A) -> anyhow::Result<T>;
+}
+
+impl<P> StateMachine<AttestationEngine<P, Attesting>, KeyNegotiator>
```

Comment:
Do we have any tests for this crate?

Reply:
Since it's a refactoring, I decided to add tests as a separate PR. So that all abstractions a "code-reviewed" first.